### ISSUE TITLE
Mutual dependency on modules results in error on some test executables with custom runner

### ISSUE THREAD

**Issue by davidgm94** [ziglang/zig#14667]
### Zig Version

0.11.0-dev.1639+438b71155

### Steps to Reproduce and Observed Behavior

I am sorry I can't provide a minimal repro, I have tried with a couple of files but it built fine..

```
git clone https:/github.com/davidgm94/rise.git
cd rise
git checkout zig-test-module-dependency-error
zig build
```

Observe that executables compile just fine, but a test executable doesn't. You can disable building the problematic test executable by commenting out line 49:
```zig
            const cpu_driver_test = try createCPUDriver(builder, modules, architecture, true);
```
The error messages:
```zig
src/bootloader.zig:6:21: error: no package named 'lib' available within package 'bootloader'
const lib = @import("lib");
                    ^~~~~
referenced by:
    Information: src/bootloader.zig:39:15
    Information: src/bootloader.zig:34:32
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

src/bootloader/limine/limine.zig:378:28: error: no package named 'privileged' available within package 'bootloader'
const privileged = @import("privileged");
                           ^~~~~~~~~~~~
src/bootloader/limine/limine.zig:372:21: error: no package named 'lib' available within package 'bootloader'
const lib = @import("lib");
                    ^~~~~
error: cpu_driver_test_x86_64...
error: The following command exited with error code 1:
/home/david/zig/0.11.0-dev.1639+438b71155/files/zig test --entry entryPoint /home/david/dev/rise/src/cpu/test.zig --test-runner /home/david/dev/rise/src/cpu/entry_point.zig -fno-strip --cache-dir /home/david/dev/rise/zig-cache --global-cache-dir /home/david/.cache/zig --name cpu_driver_test_x86_64 -fno-stack-check -fno-stack-protector -mno-red-zone -fno-omit-frame-pointer -mcmodel kernel -target x86_64-freestanding-none -mcpu x86_64-mmx+soft_float-sse-sse2-x87 --script /home/david/dev/rise/src/cpu/arch/x86/64/linker_script.ld --test-no-exec --pkg-begin lib /home/david/dev/rise/src/lib.zig --pkg-begin lib /home/david/dev/rise/src/lib.zig --pkg-end --pkg-end --pkg-begin bootloader /home/david/dev/rise/src/bootloader.zig --pkg-begin bootloader /home/david/dev/rise/src/bootloader.zig --pkg-end --pkg-begin lib /home/david/dev/rise/src/lib.zig --pkg-end --pkg-begin privileged /home/david/dev/rise/src/privileged.zig --pkg-end --pkg-end --pkg-begin privileged /home/david/dev/rise/src/privileged.zig --pkg-begin privileged /home/david/dev/rise/src/privileged.zig --pkg-end --pkg-begin lib /home/david/dev/rise/src/lib.zig --pkg-end --pkg-begin bootloader /home/david/dev/rise/src/bootloader.zig --pkg-end --pkg-end --pkg-begin cpu /home/david/dev/rise/src/cpu.zig --pkg-begin cpu /home/david/dev/rise/src/cpu.zig --pkg-end --pkg-begin lib /home/david/dev/rise/src/lib.zig --pkg-end --pkg-begin bootloader /home/david/dev/rise/src/bootloader.zig --pkg-end --pkg-begin privileged /home/david/dev/rise/src/privileged.zig --pkg-end --pkg-end --main-pkg-path /home/david/dev/rise/src -fPIC --enable-cache 
error: the following build command failed with exit code 1:
/home/david/dev/rise/zig-cache/o/49a960f0769f772a19c8d5764153e04b/build /home/david/zig/0.11.0-dev.1639+438b71155/files/zig /home/david/dev/rise /home/david/dev/rise/zig-cache /home/david/.cache/zig
```

### Expected Behavior

Everything builds fine

**Comment by davidgm94** [comment]
This is related to https://github.com/ziglang/zig/pull/14588 and https://github.com/ziglang/zig/issues/13970
Maybe a subtle thing?

**Comment by davidgm94** [comment]
If this is my error, please do tell me, because I am not so confident with the new build API for modules

**Comment by mlugg** [comment]
TL;DR: resolved by #14664

It's actually possible for it to fail with the error you're seeing on the non-test build, just seemingly less likely (it's probably related to cache population, since it happened on my initial clone). This issue is a little bit your fault, but mainly the compiler's.

Basically, the `zig build-exe` interface (which `zig build` is calling under the hood) currently doesn't have a way to specify recursive dependencies - they have to form a strict tree. If you were forming an actual dependency chain, you'd probably just segfault the build runner, but it seems like you're not currently doing that - you're probably duplicating modules somewhere accidentally. I haven't looked too closely at your `build.zig` logic, but dissecting the `build-exe` commands (you can see them through `zig build --verbose`), I can see that the actual module structures being used look like this (excuse the poor jspaint diagram):
![image](https://user-images.githubusercontent.com/7289241/219803717-5a8cefd5-078f-44a6-a7d0-85451b540b40.png)
The key thing here is that all the modules are duplicated a bunch, rather than referring to the same thing. The dependency structure you *want* for `disk_image_builder` probably looks like this:
![image](https://user-images.githubusercontent.com/7289241/219804828-1e389193-2cab-475d-87a0-97d67cd40f83.png)
I can't be bothered to draw the one for `entry_point`, but you get the idea.
This probably is something you need to resolve in your `build.zig`, but it won't be able to work properly until #14664 gets in. The weird inconsistent errors are because the structure given here *shouldn't* really work at all, but it almost-sorta-does because of some weird bugs in the compiler. That PR also fixes those issues, so using this buggy structure consistently gives you a more helpful error:
```
/home/mlugg/rise/src/lib.zig:1:1: error: file exists in multiple packages
/home/mlugg/rise/src/lib.zig:1:1: note: root of package root.lib
/home/mlugg/rise/src/lib.zig:1:1: note: root of package root.lib.lib
/home/mlugg/rise/src/lib/filesystem.zig:1:21: note: imported from package root.lib
const lib = @import("../lib.zig");
                    ^~~~~~~~~~~~
/home/mlugg/rise/src/lib/filesystem/fat32.zig:3:21: note: imported from package root.lib
const lib = @import("../../lib.zig");
                    ^~~~~~~~~~~~~~~
/home/mlugg/rise/src/bootloader.zig:1:1: error: file exists in multiple packages
/home/mlugg/rise/src/bootloader.zig:1:1: note: root of package root.bootloader
/home/mlugg/rise/src/bootloader.zig:1:1: note: root of package root.bootloader.bootloader
```

**Comment by davidgm94** [comment]
I have applied the PR and can confirm that it's working! @mlugg was so kind to reach out to me in Discord and guide me through on how to solve the issue. I appreciate the help, many thanks!
